### ISSUE TITLE
space.xhtml bug in split page mode

### ISSUE THREAD

**Issue by ssbroad** [celogeek/go-comic-converter#34]
Sorry, this problem is not solved. I mistakenly clicked 'completed'.

Epub generated in split page mode always produces duplicate space.xhtml. 
For example.

![xhtml bug1](https://github.com/celogeek/go-comic-converter/assets/11014064/229ec7e3-5d68-47d9-b6d4-493907b6810b)

![xhtml bug2](https://github.com/celogeek/go-comic-converter/assets/11014064/dd6482a8-9506-4212-9cce-a48ec857ba96)

I don't know why duplicate xhtml is packaged, maybe there is no problem with epub, but when you using the kindlegen to convert to mobi, it will report an error in the duplicate place:
Error(xmlmake):E27012: Item or process id already used: space_xx

When I use 7-zip.exe to delete all space.xhtml and all related statements in content.opf, epub can finally be properly converted to mobi.

What is the use of these space.xhtml? These seem to be blank pages.

I tried a lot of parameter combinations, and finally determined that the only factor affecting the emergence of this bug is '-autosplitdoublepage'.

`go-comic-converter -input "manga.zip" -output "manga.epub" -profile KS -autosplitdoublepage`

As long as split page is enabled, epub will always generate several duplicate space_xx.xhtml. 

I found that 2.69 and earlier versions did not have this problem.

Specifically, it should be the bug generated after adding crop limt. I tried to disable crop in 2.70 / 2.71, but the problem still exists.

`go-comic-converter -input "manga.zip" -output "manga.epub" -profile KS -autosplitdoublepage -crop=false`

**Comment by ssbroad** [comment]
Well, this bug doesn't always generate duplicate space.xhtml, and some comics don't have this problem. 
But our problems are still unsolved. 
I noticed that non-split mode only generates one space.xhtml at the end, while split mode generates a space.xhtml for each page that needs to be splitted. 
This problem still exists in versions 2.69 and earlier, but it does not affect reading until 2.7. 
I wonder why we need to add such a blank page (space.xhtml) at the end of the comic.

**Comment by ssbroad** [comment]
I deleted the call to the writeBlank method. 
`if err := e.writeBlank(wz, img); err != nil`
Now the program no longer generates a large number of space.xhtml. 
Is this method to remind us that there is something wrong with picture processing?

**Comment by celogeek** [comment]
Hi,
The spaces are use for landscape mode. all spaces are hidden in portrait. You can for example remove them all with "-portrait-only" mode.

When you are on landscape, each page need to have a right and a left page.

At the end of the comic, if I have only 1 left or right, then I need the other page with will be a space.
For splitted double page, I use the "center" mode for the double page if you keep it. But I need to add a space before if needed to complete the left/right previous page.

I notice bugs appears in some cases with the version 2.70. This version better manage cropping. For example if you have a double page, but one of them is actually a blank page, the previously the cropping wasn't working properly. Now it consider this as a single page instead.

But it lead to some trouble sometimes like you manage to get it.
I need to have an example to trouble shoot.

The solution I will start with, is to handle spaces differently. Instead of linked it to a specific page, I will create a uniq spaces with an increment. No matter where I need it, it won't be a duplicate.
Unfortunately each page need to be used only once in an ePub. In case of spaces, it would have been perfect to reuse the same page again and again.

**Comment by celogeek** [comment]
Can you share the file that cause trouble and your options ?
```
go-comic-converter -show
```

I try a lot of combinaison and different file case, but I can't generated the issue. I will keep trying, but it will be easier to have an example to play with.

**Comment by ssbroad** [comment]
I uploaded two representative files. 
The original file is very large, so I deleted some double pages, but these simplified files are consistent with the errors generated by the original file.


Let me elaborate on what I found:

KRR_01.zip https://pixeldrain.com/u/eB9JvtD1

No duplicate space.xhtml:  `go-comic-converter -input "KRR_01.zip" -output "KRR_01.epub" -profile KS -manga -autosplitdoublepage -keepdoublepageifsplit=false -titlepage 0 -noblankimage=false`

Has duplicate space.xhtml: `go-comic-converter -input "KRR_01.zip" -output "KRR_01.epub" -profile KS -manga -autosplitdoublepage -keepdoublepageifsplit=false -titlepage 0 -noblankimage=false -hascover=false`

![xhtml bug2](https://github.com/celogeek/go-comic-converter/assets/11014064/b6cd7dee-4686-4379-a84e-4cdbd1a55c43)


The parameter that caused the error: `-hascover=false`




With You and the Rain vol.1.zip https://pixeldrain.com/u/dZzFupTB

Has duplicate space.xhtml: `go-comic-converter -input "With You and the Rain vol.1.zip" -output "With You and the Rain vol.1.epub" -profile KS -manga -autosplitdoublepage`

![xhtml bug3](https://github.com/celogeek/go-comic-converter/assets/11014064/d7af9aef-1cdf-4a86-aafa-26ca949b997e)

An error will be generated as long as the _autosplitdoublepage_ parameter is added

Regardless of whether there are duplicate xhtml or not, as long as autosplitdoublepage is enabled, the comic always generates a large number of blank pages at the end, which are useless when reading.

![xhtml bug4](https://github.com/celogeek/go-comic-converter/assets/11014064/23b467d6-d32b-498b-a627-e419a8c43a2a)

I deleted all the code about blank pages in epub.go and epub_templates_content.go, and it feels good so far.

**Comment by celogeek** [comment]
Thanks a lot for the sample. It will help me to figure out what is the matter with the spaces.

Remove those line of code will just make the manga failed to properly display when you activate the landscape, specifically when double page is involved.

If you don't care about the landscape, then you can use the "-portrait-only" option. You won't have any spaces with this option.

The current implementation make it difficult to coordinate the TOC/Content, with the files. Both a generated separately.
It make be easier to first populate the images with the necessary spaces, and then record the result into the TOC afterward.

I will see if I can figure out a quick fix, or if I need to rewrite the processing.

**Comment by celogeek** [comment]
I've found the issue.

You have some double page that result into another double page for each cropped part.
It's when you have a very large picture like a banner.
After cutting in half, each part are still wider (width > height).

The previous code mark then as double page, and then add spaces there.
I fix the issue

please try this:
```
go install go install github.com/celogeek/go-comic-converter/v2@f7f8680eb3e0d18d822a4b3ed07dbc03465f8bb0
```

It's the branch 34-spacexhtml-bug-in-split-page-mode

I will deal with another issue on with the crop-limit params. It block the detection of empty pages. You have an example with a double page which is actually a simple page as the right side is entirely blank.

**Comment by ssbroad** [comment]
Thank you for your explanation, I just found out the landscape mode of Kindle! 
This is very useful! 

But after my tests, the comicss generated by splitdoublepage do not match the screen size in landscape mode. 
Comics generated in single-page (disable splitdoublepage) can normally match the screen. 

Why?

**2 pics from non-split**
![upscreenshot1](https://github.com/celogeek/go-comic-converter/assets/11014064/03b1f22b-4fe1-4a41-a19a-f7bca972541d)


**2 pics from splitdoublepage**
![upscreenshot2](https://github.com/celogeek/go-comic-converter/assets/11014064/d80e305a-d77c-43ec-8204-c8de8abb6fa5)

The picture comes from a screenshot of Kindle.

**Comment by celogeek** [comment]
When you split the double page, it will cut the image in 2.
On portrait you will see 1 then the other.
On landscape, you will see one on the left and another on the right. The end result should be similar to a doublepage shown in the middle.
I fix the issue.
The problem here, is that each split is similar to a double page too (width > height)

**Comment by celogeek** [comment]
Try this patch:

```
go install go install github.com/celogeek/go-comic-converter/v2@f9ebe71b9e4d97d540f71ddb010d9aad6f73be87
```

I move the split of double page after autocropping. This way the render is working on landscape.

**Comment by ssbroad** [comment]
`go install github.com/celogeek/go-comic-converter/v2@f9ebe71b9e4d97d540f71ddb010d9aad6f73be87`

This version has the same problem as 2.6.9, that is, it cannot accurately crop double page.

https://github.com/celogeek/go-comic-converter/issues/28#issuecomment-1913102884

**Comment by celogeek** [comment]
The issue is when you crop a double page, you have to crop it as a 1 big image that need to be cut in 2.
So each side of the crop need to have the same size exactly otherwise it's not possible to place them side by side in landscape and have a similar view like when you use the doublepage uncut.

I only see the "portrait-only" mode, when this cut can be different on each side, as you don't have to place them side by side. I was thinking of having 2 differents way to crop the splitted images, based on this option.

I can also may be add an option to indicate if we need to consider doublepage as 1 image or 2 differents images.
based on the source, it may be one or the other I guess. most of the time doublepage is 1 big image.
The case when it's 2 differents pages, it's like you scan 2 page at once.

**Comment by ssbroad** [comment]
Did you see the comment I deleted? 
Yes, the problem is, as you said, we can only choose between a perfect landscape display and accurately cropping.

**Comment by celogeek** [comment]
I receive an email with the comment ;) yes. and it's interesting to have a way to deal with the 2 cases.

**Comment by celogeek** [comment]
I think if we want to satisfy all cases we can:
- portrait-only: cut in 2, auto crop each side (we don't care to display side by side)
- landscape: have an option like -doublepage-is-two-separate-image
  - if this option is on, we cut and autocrop, like we can do with 2 differents images. Eventually we can even separate them like a 2 images in 1.
  - if this option is off, we first crop then cut like the current patch to have a proper render in landscape

**Comment by ssbroad** [comment]
~~Well, I personally prefer to turn on -doublepage-is-two-separate-image.~~ 

I think both options are good

By the way, let me talk about why landscape mode is indispensable for kindle manga. 

When splitting double pages, we will encounter a problem that is inevitable anyway. Some pages should not be split.
![upscreenshot1](https://github.com/celogeek/go-comic-converter/assets/11014064/03b1f22b-4fe1-4a41-a19a-f7bca972541d)
Unless we sift out these pictures manually in advance and deal with them separately, but this is obviously not what we want (I don't want to be spoiled before reading comics).
landscape mode can help us restore these pages that should not have been segmented.


The kindle device is really not friendly for comics. I think this is why some comic lovers want to jailbreak their kindle.

At present, I will first use go-comic to convert comics to epub, and then convert them to mobi format through kindlegen.

Of course, we can also send the epub to Amazon and wait for the network to push it.

To be honest, the process is a bit tedious.

~~Not long ago I discovered the [MOBI](https://github.com/leotaku/mobi) package built using the go language. I wonder if it is possible to introduce go-mobi into go-comic and directly output azw3?~~


It seems that you have examined the problem of directly outputting azw3 to break through the resolution limit.

**Comment by ssbroad** [comment]
OMG, I came up with a less than perfect solution

Let’s see what these pictures that shouldn’t be splitted have in common.
![Trigun Maximum v01 (2004) (Digital) (Lovag) 008](https://github.com/celogeek/go-comic-converter/assets/11014064/d9406206-884d-4ba6-a09b-f2c17063c584)
![Dr2](https://github.com/celogeek/go-comic-converter/assets/11014064/303ff8b7-e844-4211-9717-46b717dbe6d0)
![Dr](https://github.com/celogeek/go-comic-converter/assets/11014064/b936f8d9-392d-4b82-82dc-aab0907ff517)

**The middle area of ​​these images is not blank! Type A**

Most of the double-page images that should be split are blank in the middle!  Type B

![Rozen-Maiden_04_016](https://github.com/celogeek/go-comic-converter/assets/11014064/80f96ad6-a22a-4708-913e-d1568193a259)



Perhaps we can judge the average grayscale of the middle part of a picture in split mode. 

If the value is close to 1 (indicating white), then the area is roughly white. Then cut and autocrop.

If the value is less than 0.99 or 0.98, then the image is type A

For type A, here are 2 ways to deal with it

1. Rotate the image 90 degrees without split, then crop
2. crop then cut

There are several obvious problems with this approach
1. will increase the time of image processing
2. how to control the range of the middle detection area
3. the minimum value that determines whether the area is white

The above are all my random ideas, which will make the program more complicated and difficult to use. Maybe I should try to learn go?^_^

**Comment by ssbroad** [comment]
The reason why pictures cannot match the screen normally in landscape mode still seems to have not been found.

unable to match - KRR_16.zip https://pixeldrain.com/u/k5WffqMM

match - KRR_17.zip https://pixeldrain.com/u/ekzzTrXX

The number and resolution of pictures in the two files above are exactly the same. My set `-profile KS -manga -titlepage 0 -noblankimage=false -hascover=false -sort 2 -autosplitdoublepage -keepdoublepageifsplit=false`

And this version does not solve the problem of xhtml duplication

`go install github.com/celogeek/go-comic-converter/v2@f9ebe71b9e4d97d540f71ddb010d9aad6f73be87`

space.xhtml bug - KRR_17-2.zip https://pixeldrain.com/u/4dsmud66

**Comment by celogeek** [comment]
Can you give me the output of the "-show" option.
If not already the case, can you set the aspect ratio to 1:1.6
```
-aspect-ratio 1.6
```

As to detect if the double image is actually 2 differents images or 1 single big one, the autodetection with a blank separation in the middle is interesting. But the example you give me is the exception. You have a white band, and then it will be consider as 2 differents images.
Also no matter the type of image, having it spread properly on 1 screen is interesting anyway.

In your KRR_17 file, the up4 is a very large image. It could be split into 3 to 4 pieces actually.
If not, then each side will be tiny in the middle of the screen.

Which kindle do you have ? which version ? I do my test on a kindle scribe. I don't see the point of landscape mode on a smaller device.

**Comment by ssbroad** [comment]
I didn't create any settings file, I entered the parameters manually in cmd every time

My device is ks, same as yours.

Did you test that both files can be displayed normally in landscape mode? (using the same settings as mine)

KRR_16.zip KRR_17.zip

**Comment by ssbroad** [comment]
It looks like adding the aspect-ratio is the right thing to do, I will continue testing

**Comment by celogeek** [comment]
I introduce the aspect ratio in order to support ePub on tablet with colors.
Kindle best aspect ratio no matter of the device is 1.6.
You can also use 1.55 if you are greedy ;) 

I've fix an issue with the crop limit to detect empty page.
I see you have half of an image which are empty, on a double page. But the crop limit forbid to remove it.
I wonder if I should detect that half of the double image is empty and cut it.

the patch is 27d617640a1e1df3bb1469911253e1d3fda83870

```
go install go install github.com/celogeek/go-comic-converter/v2@27d617640a1e1df3bb1469911253e1d3fda83870
```

**Comment by celogeek** [comment]
For landscape mode support, I have to keep the system as it is now.
But for portrait only we can:
- cut then crop as we don't need to display both side on the same page
- rotate image if the cut result is wider than larger

I can't do rotation on landscape, the side by side rendering will be wrong. or I can disable side by side rendering if the result is a wider image...

looking for a compromise

**Comment by ssbroad** [comment]
Are you referring to this picture?

![upup02-3](https://github.com/celogeek/go-comic-converter/assets/11014064/bcccd6e9-4e24-4cb3-8e0c-fa9d7ad2deb5)


I don't quite understand what the problem would be, since I've always chosen to `-noblankimage=false`

In my opinion blank pages are very rare in comics, and the very few blank pages will not have any impact on the reading experience.

I think keeping blank pages can ensure that the correct double page is restored in landscape.

Thanks to your work, I decided to use the [version](https://github.com/celogeek/go-comic-converter/issues/34#issuecomment-2096302533) below as my workhorse because I prefer precise cut.

`go install github.com/celogeek/go-comic-converter/v2@f7f8680eb3e0d18d822a4b3ed07dbc03465f8bb0`

and with `-aspect-ratio 1.6`

**Comment by celogeek** [comment]
Can you try this one 62fc520ebb2463391b39121acd873f8a303ef3c4 ?

Play with portrait-only option.
When activated, the cut will be done also on each side of the double page splitted.
When disable, as we need to preserve aspect ration of the splitted double page in landscape mode, I will crop before cut.

**Comment by ssbroad** [comment]
I have done some new tests on 62fc520

After turning on portrait-only, we can more accurately crop double-page images.

![double_p1_portrait-only](https://github.com/celogeek/go-comic-converter/assets/11014064/a622ea4a-e66b-4f06-a2c8-e1b190a0ddf4)

After turning off portrait-only, we get 2 images of the same height, at the expense of the first image still leaving a lot of white space.

![double_p1](https://github.com/celogeek/go-comic-converter/assets/11014064/8eceb36f-7c25-4b95-944b-e04fde0e9f2a)

However, the display effect of the first picture (from turning off portrait-only) in KS is not bad, and the blank part has been reduced, although it is still a little short of precision cropping.

![screenshot_KS](https://github.com/celogeek/go-comic-converter/assets/11014064/e77d5804-6c1e-4ba9-bfa5-f16171ec03ad)

**Comment by ssbroad** [comment]
I need landscape mode, but I only turn it on for large images (Type A) and then turn it off immediately.

![Dr2](https://github.com/celogeek/go-comic-converter/assets/11014064/303ff8b7-e844-4211-9717-46b717dbe6d0)

For my reading habits (only turning on horizontal screen once when encountering a large picture), the benefits brought by precise cropping are obviously better. A double-page comic may have only 1-3 large images, or sometimes none at all.

I won't turn on portrait-only option, I need landscape mode to deal with potentially large images. But I also wouldn't compromise on precise cropping, as there are very few large images.

Everyone’s reading habits are different, the above is just my personal opinion

**Comment by celogeek** [comment]
I add the option you want here 87faac877cfb3475e7aa673d92b7cd0965aaab7a

```
-keepsplitdoublepageaspect=0
```

It's on by default, and force on or off with some parameters (portrait only or apple compat)

In you case, if you mostly read in portrait, I suggest those options:
```
-portrait-only
-autosplitdoublepage=0
-autorotate
```

Every double page will be rotated to 90 degrees, so you can simple turn your kindle without activating or disabling landscape option.
You will have a fullscreen image without cut. I usually disable the autosplit because I don't need it on a kindle scribe.

**Comment by ssbroad** [comment]
keepsplitdoublepageaspect is exactly what I want, thank you again for your efforts


Now my options:
the main body of comics is mainly double pages (may contain a few single pages, such as a cover)
`-profile KS -manga -titlepage 0 -noblankimage=0 -sort 2 -aspect-ratio 1.6 -hascover=false -autosplitdoublepage -keepdoublepageifsplit=0 -keepsplitdoublepageaspect=0 -crop-limit 12`

the main body of comics is mainly single pages
`-profile KS -manga -titlepage 0 -noblankimage=0 -sort 2 -aspect-ratio 1.6 -hascover=false -autorotate -crop-limit 12`


Finally allow me to submit a coding bug

When the program is processing a compressed package containing non-English folders, the progress information at the bottom after opening the comic will be garbled.

no fold : https://pixeldrain.com/u/4xUqyhxR
non-english fold https://pixeldrain.com/u/zNucPhri

![upscreenshot](https://github.com/celogeek/go-comic-converter/assets/11014064/68e8f90a-c4f2-4d5a-a37e-da62e0ae54ec)


Compressed and non-compressed packages that do not contain folders do not have this problem.

**Comment by celogeek** [comment]
I think this is another issues. Can you close this one if that's ok, and create a new one with this specific bug ?
I've tooks the files, I will take a look. (not sure to get the diff between the 2 of them)
